airflow/plugins/operators/littlepay_raw_sync.py

```diff
@@ -0,0 +1,217 @@
+from typing import ClassVar, Dict, List
+
+import boto3
+import pendulum
+from calitp_data_infra.auth import get_secret_by_name
+from calitp_data_infra.storage import (
+    PARTITIONED_ARTIFACT_METADATA_KEY,
+    PartitionedGCSArtifact,
+    get_fs,
+    get_latest_file,
+)
+from pydantic.class_validators import validator
+from pydantic.error_wrappers import ValidationError
+from pydantic.main import BaseModel
+from tqdm import tqdm
+from tqdm.contrib.logging import logging_redirect_tqdm
+
+from airflow.models import BaseOperator
+
+LITTLEPAY_RAW_BUCKET = os.getenv("CALITP_BUCKET__LITTLEPAY_RAW")
+
+
+class LittlepayFileKey(str):
+    @classmethod
+    def __get_validators__(cls):
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
+        assert len(v.split("/")) == 3 and v.endswith(".psv")
+        return cls(v)
+
+    @property
+    def agency(self) -> str:
+        return self.split("/")[0]
+
+    @property
+    def entity(self) -> str:
+        return self.split("/")[1]
+
+    @property
+    def filename(self) -> str:
+        return self.split("/")[2]
+
+
+class LittlepayS3Object(BaseModel):
+    Key: LittlepayFileKey
+    LastModified: pendulum.DateTime
+    ETag: str
+    Size: int
+    StorageClass: str
+
+
+class RawLittlepayFileExtract(PartitionedGCSArtifact):
+    bucket: ClassVar[str] = LITTLEPAY_RAW_BUCKET
+    partition_names: ClassVar[List[str]] = ["instance", "filename", "ts"]
+    instance: str
+    filename: str
+    ts: pendulum.DateTime
+    s3bucket: str
+    s3object: LittlepayS3Object
+
+    @validator("ts", allow_reuse=True)
+    def coerce_ts(cls, v):
+        if isinstance(v, datetime):
+            return pendulum.instance(v)
+        return v
+
+    @property
+    def table(self) -> str:
+        return self.s3object.Key.entity
+
+    # TODO: why does this not override the parent model field?
+    # @property
+    # def filename(self) -> str:
+    #     return self.s3object.Key.filename
+
+
+def sync_file(src_bucket: str, file: RawLittlepayFileExtract, s3client, fs):
+    try:
+        # TODO: this kinda overlaps with get_latest()
+        fileinfo = get_latest_file(
+            bucket=file.bucket,
+            table=file.table,
+            prefix_partitions={
+                "instance": file.instance,
+                "filename": file.filename,
+            },
+            partition_types={
+                "ts": pendulum.DateTime,
+            },
+        )
+        try:
+            metadata_str = fs.getxattr(
+                path=f"gs://{fileinfo.name}", attr=PARTITIONED_ARTIFACT_METADATA_KEY
+            )
+        except KeyError:
+            print(f"metadata missing on {fileinfo.name}")
+            raise
+        prior = RawLittlepayFileExtract(**json.loads(metadata_str))
+        save = (
+            file.s3object.LastModified != prior.s3object.LastModified
+            or file.s3object.ETag != prior.s3object.ETag
+        )
+    except FileNotFoundError:
+        save = True
```

Comment:
What do we end up saving if we reach this, in practice? Should it go to a different place than the results of successful tries?

Reply:
This condition is hit whenever we have never synced a source file before, so it's a happy path and probably the most common path.